### PR TITLE
Improve hook

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

File: chainer/function.py [hunk 1]
```diff
@@ -189,8 +189,10 @@ def __call__(self, *inputs):
         if self.type_check_enable:
             self._check_data_type_forward(in_data)
 
-        hooks = collections.OrderedDict(chainer.get_function_hooks())
-        hooks.update(self.local_function_hooks)
+        hooks = chainer.get_function_hooks()
+        if self.n_local_function_hooks != 0:
```
Comment (unnonouno) [inline chainer/function.py]: How about using `if len(hooks) != 0:` ?

Comment (okuta) [inline chainer/function.py]: If we use `len(self.n_local_function_hooks)`, a python object is created.
It has overhead.

Comment (unnonouno) [inline chainer/function.py]: Ah, OK. I understand it. Does it need to make a **public** method?

Comment (okuta) [inline chainer/function.py]: OK, I made a private method.

**Pull request merged**